### PR TITLE
fix(lerobot_local): reconstruct OLD-FORMAT in-model normalization dropped on load

### PR DESCRIPTION
**Pull request by cagataycali**
## Problem

Loading a canonical **pre-processor-era** lerobot checkpoint through the `lerobot_local` provider runs it with **normalization dropped**. These checkpoints (the zoo a user grabs first to try `lerobot_local` -- `lerobot/act_aloha_sim_transfer_cube_human`, `diffusion_pusht`, the tdmpc/vqbet entries) store their `Normalize` modules *inside* the policy, so `model.safetensors` carries `normalize_inputs.*` / `normalize_targets.*` / `unnormalize_outputs.*` buffers and `config.json` carries a `normalization_mapping`, with **no** `policy_preprocessor.json` / `policy_postprocessor.json`.

Current lerobot no longer registers those in-model modules (normalization moved to the processor pipeline), so `PreTrainedPolicy.from_pretrained` -> `_load_as_safetensor(strict=False)` drops the 8 buffers as *unexpected keys* (only a `WARNING:root` line). With no processor JSON to replace them, `ProcessorBridge` had nothing to load and the policy ran raw: observations reached the model un-normalized **and** predicted actions reached the robot un-unnormalized.

For a `MEAN_STD` checkpoint this is not an "arm barely moves" under-motion -- raw z-scored actions applied as robot units make the arm **flail**.

## Fix

`ProcessorBridge` now reconstructs the pre/post pipelines from those same in-model buffers using lerobot's own recovery machinery -- `extract_normalization_stats` (the exact scanner `migrate_policy_normalization` uses) + the public `make_pre_post_processors` factory -- so an old-format checkpoint runs normalized with **zero user action**, equivalent to running lerobot's migration script.

The reconstruction is a third, best-effort fallback in `ProcessorBridge.from_pretrained`, added after the existing `policy_*.json` and `norm_stats.json` paths. It fires **only** when a checkpoint ships no processor configs, no `norm_stats.json`, but *does* carry in-model buffers -- modern checkpoints are untouched. It requires the policy config (threaded from the loaded policy); when the config is absent or the recovery API is unavailable it degrades to the prior passthrough so the existing missing-postprocessor diagnostic still fires. When buffers are present but reconstruction fails, it warns with the exact manual migration command instead of failing silently.

## Rollout evidence (MuJoCo, `act_aloha_sim_transfer_cube_human`)

Identical scene / home keyframe / cube / `seed=42`; only normalization differs. Right-gripper trajectory over a 400-step rollout, ground-truthed on `data.site_xpos`:

| condition | right-gripper path | right-gripper z-range | left-gripper path |
|---|---|---|---|
| **before** (normalization dropped) | 3.36 m | [0.03, **0.90**] m (flails ~90 cm up) | 1.76 m |
| **after** (reconstructed) | 0.62 m | [0.16, 0.32] m (controlled reach-down) | 0.17 m (stays home) |

Side-elevation view (world-z -> screen-vertical) of the same two rollouts -- left = after, right = before:

before vs after

Load-time log after the fix (no more misleading warning):
```
lerobot_local: lerobot/act_aloha_sim_transfer_cube_human ships no processor configs but carries OLD-FORMAT in-model normalization; reconstructed pre/post pipelines from its 3 in-model buffer group(s) ['action', 'observation.images.top', 'observation.state'] (equivalent to lerobot's migrate_policy_normalization). The policy runs normalized.
```

## Tests

`tests/policies/lerobot_local/test_in_model_normalization_fallback.py` builds a synthetic OLD-FORMAT checkpoint on disk (real `safetensors` in-model buffers + a real `ACTConfig`) and drives the real lerobot machinery -- no mocks:
- reconstruction rebuilds a postprocessor carrying the exact in-model `action` unnormalization stats;
- the fallback is skipped without a policy config (backward-compat passthrough preserved);
- a checkpoint with no in-model buffers stays a passthrough (no false positives on modern checkpoints).

All three fail before the fix (the `policy_config` path did not exist) and pass after. Full `lerobot_local` suite green; `ruff` + `mypy` clean.

## Scope

The fix corrects the action/state normalization *scale* (stops the flail). It does not by itself enable a task success in sim -- the render domain gap and per-embodiment state/gripper mapping are separate barriers -- but a policy running in its trained normalized space is the necessary first step.

## Robustness: import-guard degradation

The reconstruction imports `lerobot.policies` (for `make_pre_post_processors`) and `lerobot.processor.migrate_policy_normalization`. These can fail at **definition time**, not only with `ImportError` -- an unrelated broken sibling policy module (e.g. a dataclass with a non-default field after a default one in `lerobot.policies.groot` raises `TypeError` while importing the `lerobot.policies` package) takes down the whole import. The guard now degrades to passthrough on **any** import-time failure so an unrelated lerobot defect cannot crash an ACT/diffusion checkpoint load. The test module guard is broadened the same way (`pytest.importorskip` only catches `ImportError`).

A fourth regression test (`test_reconstruction_degrades_when_lerobot_import_raises_non_import_error`) pins this: with `lerobot.policies` import monkeypatched to raise `TypeError`, `_load_in_model_normalization_fallback` returns `(None, None)` instead of propagating. It fails on the prior `except ImportError` guard and passes after. Full `lerobot_local` suite green (489 passed); `ruff` + `mypy` clean.

---

## Round 2 changelog

- **CodeQL unused-import (2 findings, test file lines 42-43):** the two module-level lerobot probe imports were bound-name imports used only for their side effect of executing the package bodies at collection time (so the skip guard fires on any import-time failure, including a definition-time `TypeError`, not just `ImportError`). Their bound name was never referenced, which the scanner flagged and the `# noqa: F401` did not silence. Switched both to `importlib.import_module(...)`: identical import machinery (module body still executes, definition-time errors still propagate into the `except` guard and skip the module) with no unused binding. Finding cleared at source rather than suppressed. `ruff check` / `ruff format` clean, ASCII-only. (commit `6461eef`)

## lerobot API-drift fix

`make_pre_post_processors` is not a stable member of the `lerobot.policies` top-level re-export (current lerobot no longer exports it there). The deferred import now sources it from its canonical module `lerobot.policies.factory` — the exact module lerobot's own `migrate_policy_normalization` imports it from — so it resolves across lerobot releases and clears the static `lerobot`-import-surface drift guard.

### CODE REVIEW

**Review by yinsong1986 — DISMISSED** [review]
## Summary

Adds a third, best-effort fallback to `ProcessorBridge.from_pretrained` that reconstructs pre/post normalization pipelines for OLD-FORMAT (pre-processor-era) lerobot checkpoints. These canonical-zoo checkpoints (`act_aloha_sim_transfer_cube_human`, `diffusion_pusht`, tdmpc/vqbet) carry `Normalize` modules as in-model `model.safetensors` buffers that current lerobot no longer registers, so `PreTrainedPolicy.from_pretrained` silently drops them and the policy runs un-normalized (raw MEAN_STD actions -> the arm flails). The new path uses lerobot's own `extract_normalization_stats` + `make_pre_post_processors` (the machinery of `migrate_policy_normalization`) to rebuild the pipelines from those buffers, threading the loaded policy's config via a new optional `policy_config` kwarg. It fires only when a checkpoint ships no processor configs, no `norm_stats.json`, but does carry in-model buffers; modern checkpoints are untouched.

## What's good

- Additive and backward-compatible: new `policy_config: Any | None = None` kwarg defaults to skipping the fallback, so callers that never pass it keep the prior passthrough behaviour. No `__all__` change; the two new helpers stay private.
- Exception discipline holds up. `_load_checkpoint_state_dict` catches `ImportError` (no safetensors/hub) and `(HfHubHTTPError, OSError, ValueError)` for the hub path -- verified that `EntryNotFoundError` (sharded checkpoints) subclasses both `HfHubHTTPError` and `OSError`, so it degrades cleanly rather than raising into the load path. The best-effort `make_pre_post_processors` guard carries the project-standard `# noqa: BLE001` and, when buffers are present but reconstruction fails, warns with the exact manual migration command instead of failing silently.
- API wiring checked against lerobot: `make_pre_post_processors` returns a 2-tuple (matches the unpack), `dataset_stats` is a valid `ProcessorConfigKwargs` kwarg whose type matches `extract_normalization_stats`'s return, and the `from lerobot.policies import make_pre_post_processors` path is exported on lerobot main / the >=0.5.2 source install these VLA features already depend on (an older lerobot raises `ImportError`, caught and degraded to passthrough).
- Real-machinery tests (no mocks) covering reconstruction, the no-config backward-compat skip, and the no-buffers no-false-positive case; no host paths and no non-ASCII in the new strings.

## Verification suggestions

Standard `hatch run test` on the `lerobot_local` suite is sufficient. To spot-check the reported behaviour change beyond CI, run the `act_aloha_sim_transfer_cube_human` MuJoCo rollout and confirm the load-time log emits the "reconstructed pre/post pipelines from its N in-model buffer group(s)" INFO line and the right-gripper trajectory stays in the controlled z-range from the PR table.

**Review by yinsong1986 — DISMISSED** [review]
## Summary

Adds a third best-effort fallback to `ProcessorBridge.from_pretrained` that reconstructs pre/post normalization pipelines for OLD-FORMAT (pre-processor-era) lerobot checkpoints (`act_aloha_*`, `diffusion_pusht`, tdmpc/vqbet). Those checkpoints carry `Normalize` modules as in-model `model.safetensors` buffers that current lerobot no longer registers, so they were silently dropped on load and the policy ran un-normalized (MEAN_STD z-scored actions applied as robot units -> arm flails). The fix reads the checkpoint's single-file state dict, runs it through lerobot's own `extract_normalization_stats` + `make_pre_post_processors` (the migration script's machinery), and rebuilds the pipelines with zero user action. The fallback is correctly gated to fire only when both standard pipelines are absent AND no `norm_stats.json` exists AND in-model buffers are present -- modern checkpoints are untouched. The head commit itself is a narrow CodeQL cleanup (switching two side-effect probe imports in the test to `importlib.import_module`).

## What's good

- Tight gating (`preprocessor is None and postprocessor is None and policy_config is not None`) means the fallback cannot fabricate a pipeline for a modern checkpoint or clobber an already-loaded one.
- Every failure path degrades to `(None, None)` passthrough so the existing missing-postprocessor diagnostic still fires; the `except Exception` on the import guard is justified (an unrelated broken sibling policy module can fail at definition time, not just `ImportError`) and matches the documented design.
- New `policy_config` kwarg is additive with a `None` default -- backward compatible, no one-way door.
- Real-machinery tests (synthetic on-disk safetensors + real `ACTConfig`, no mocks), ASCII-clean strings, no host paths, no private names in `__all__`. The prior CodeQL unused-import finding is resolved at source rather than suppressed.

**Review by yinsong1986 — APPROVED** [review]
## Summary

Adds a third, best-effort fallback to `ProcessorBridge.from_pretrained` that reconstructs the pre/post normalization pipelines for OLD-FORMAT (pre-processor-era) lerobot checkpoints. Those canonical-zoo checkpoints (`act_aloha_sim_transfer_cube_human`, `diffusion_pusht`, tdmpc/vqbet) carry their `Normalize` modules as in-model `model.safetensors` buffers plus a `normalization_mapping` in `config.json`, with no `policy_*.json`. Current lerobot no longer registers those modules, so `PreTrainedPolicy.from_pretrained` drops the buffers as unexpected keys and the policy previously ran with normalization dropped -- observations reached the model raw and actions reached the robot un-unnormalized (a MEAN_STD checkpoint flails). The fix threads the loaded policy's `config` into the bridge and, when a checkpoint ships no processor configs and no `norm_stats.json` but does carry in-model buffers, reconstructs the pipelines using lerobot's own `extract_normalization_stats` + `make_pre_post_processors` (the same machinery as `migrate_policy_normalization`).

## What's good

- **Scope discipline**: the new `policy_config` kwarg defaults to `None`, so the fallback is opt-in and modern checkpoints / existing callers are untouched; backward compatibility is preserved.
- **Fail-safe by construction**: every failure path (missing recovery API, no single-file weights, no in-model buffers, factory failure) degrades to `(None, None)` passthrough so the existing missing-postprocessor diagnostic still fires -- no new uncaught exception escapes into the load path, and `policy._load_processor_bridge` already tolerates a passthrough bridge.
- **Correct-against-installed API**: verified `make_pre_post_processors(policy_cfg=..., dataset_stats=...)` and `extract_normalization_stats` resolve against the installed lerobot; sourcing the factory from the canonical `lerobot.policies.factory` module (rather than the unstable top-level re-export) is the right drift-hardening choice.
- **Broad import guard is deliberate and documented**: catching non-`ImportError` definition-time failures (e.g. a broken sibling policy dataclass raising `TypeError`) is exactly the right call for a best-effort recovery path, and it carries an explaining `# noqa: BLE001` comment.
- **Real-machinery tests, no mocks**: the regression suite builds a synthetic OLD-FORMAT checkpoint on disk and drives the actual lerobot factory, including a pin for the definition-time-import degradation.

## Verification suggestions

Standard CI (`hatch run test` for `tests/policies/lerobot_local`, `ruff`, `mypy`) covers this. For a deeper spot-check, load a real OLD-FORMAT zoo checkpoint (`lerobot/act_aloha_sim_transfer_cube_human`) through `lerobot_local` and confirm the load-time log reports the reconstructed in-model buffer groups rather than the prior misleading missing-processor warning.
